decentraland/unity-renderer
# fix: teleport memory handling

## What does this PR change?

This PR add small tweaks to how we free memory when teleporting.

* When clearing entities in "immediate mode" now the pool is cleaned up.
* When the rendering is deactivated to start a teleport, `ParcelScenesCleaner` pending entities are cleaned up and `Resources.UnloadUnusedAssets` is called. (before this happened when the rendering was **activated**, leading to potential unwanted memory spikes)
* Added a few refactors to improve `MemoryManager` and `ParcelScenesCleaner` encapsulation. `MemoryManager` now just exposes a `OnCriticalMemory` event. `MemoryManager` responsibility was simplified, and now it doesn't uses `ParcelScenesCleaner` anymore. The `MemoryManager` responsibility of cleaning pools is still in place, but the pool cleaning code was moved to `PoolManager`.

These changes shouldn't fix the recurring memory out of bounds situation. The memory issue is a combination of many issues that are going to be addressed in subsequent PRs. Those PR are going to take advantage of the changes presented on this one.

## How to test the changes?

- Try to teleport a few times around the world.
- As this PR introduced a more aggresive unloading on teleports, we should ensure loading times don't increase very much.

## Our Code Review Standards

https://github.com/decentraland/unity-renderer/blob/master/docs/code-review-standards.md
